src/base/Assure.cc

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (C) 1996-2022 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "base/Assure.h"
+#include "base/TextException.h"
+#include "Debug.h"
+
+[[ noreturn ]] void
+ReportAndThrow_(const int debugLevel, const char *description, const SourceLocation &location)
+{
+    const TextException ex(description, location);
+    const auto label = debugLevel <= DBG_IMPORTANT ?
+                       "ERROR: Squid BUG: assurance failed" : "check failed";
+    // TODO: Consider also printing the number of BUGs reported so far. It would
+    // require GC, but we could even print the number of same-location reports.
+    debugs(0, debugLevel, label << ": " << ex);
```

Comment:
> > all Assure() failures are bugs
> 
> True, but most cache.log readers cannot tell whether the logged `ERROR` is the result of an Assure() failure.

False, all Assure()'s state "assurance failure" other log entries do not.

> And most readers do not (and should not) know what Assure() is or what special meaning we assign to "assurance".

... a problem introduced by this PR creating a novel alternative to assert. Which users are already familiar with.


> They do (or should) know what a _bug_ is though.

They know that "ERROR" means some sort of a bug.

> Since these code logic check failures are reported using non-standard "assurance" terminology, we should add a standard term like "bug" to explain what is going on.
> 

No. We should document what "assurance failure" means **somewhere**. So admin seeing it for the first time can look it up. That would be the wiki page documenting cache.log ERROR meanings.

That somewhere should not be polluting every related log line. Like I said, we already have "ERROR" to indicate type+severity of the problem.

For example, I bet you can tell me whether this is a bug or not
 ```error:14094438:SSL routines:ssl3_read_bytes:tlsv1 alert internal error (http_request_failed)```

yet nowhere does the word "BUG" appear in that log entry.

Reply:
> > They know that "ERROR" means some sort of a bug.
> 
> ERROR is usually not a bug. Very few ERROR messages are some sort of a bug.
> 

That itself would be a bug / badly written message.  From the FAQ definition:  ```a serious problem which has broken an individual client transaction and may have some effect on other clients indirectly. But has not completely aborted all traffic service```

Problems which are not internal errors for Squid, eg caused by external things should be WARNING (or just notices). FATAL and ERROR are for issues which have killed all or part of Squid service - aka "internal errors".

Reply:
> > ERROR is usually not a bug. Very few ERROR messages are some sort of a bug.
 
> That itself would be a bug / badly written message. 

I see no reason why it would but that is because your definition of `ERROR` is different from mine.

> From the FAQ definition: `a serious problem which has broken an individual client transaction and may have some effect on other clients indirectly. But has not completely aborted all traffic service`

The quoted phrase does not contradict what I said AFAICT.

N.B. Please do not misinterpret this response as an implicit agreement that we should be using the above "FAQ definition". I am just saying that even if we do use it, my statement remains correct.


> Problems which are not internal errors for Squid (eg caused by external things) should be WARNING or just notices. FATAL and ERROR are for issues which have killed all or part of Squid service - aka "internal errors".

I (and the majority of Squid code) disagree with that opinion.

Errors are things that went wrong. Warnings are notices, advice, or predictions about something that may go wrong in the future. These are natural, widely-understood concepts. There is no need to redefine them.

We should be using ERROR and FATAL labels for non-fatal and fatal (to the current process) errors. We should be using WARNING for warnings.